File: Data Structures & Algorithms/binary-search/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0;
        int end = nums.size() - 1; 

        while (start <= end) { 
            int middle = (start + end) / 2; 

            if (target < nums[middle]) {
                end = middle - 1; 
            }
            else if (target > nums[middle]) {
                start = middle + 1; 
            }
            else {
                return middle; 
            }
        }
        return -1;
    }
};

File: Data Structures & Algorithms/maximum-subarray/submission-2.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # dp[i] = maximum sum ending at index i 
        dp = [float('-inf')] * len(nums)
        dp[0] = nums[0]

        for i in range(1, len(nums)): 
            dp[i] = max(
                nums[i] + dp[i - 1], # continue curr subarray
                nums[i], # start new subarray 
            )
        
        return max(dp) 
         



        